src/schema.rs

```diff
@@ -548,6 +574,35 @@ mod impls {
         Ok(values)
     }
 
+    fn deserial_string<R: Read>(source: &mut R, size_len: &SizeLength) -> ParseResult<String> {
+        let len = deserial_length(source, size_len)?;
+        // we are doing this case analysis so that we have a fast path for safe,
+        // most common, lengths, and a slower one longer ones.
+        if len <= MAX_PREALLOCATED_CAPACITY {
+            let mut bytes = vec![0u8; len];
+            source.read_exact(&mut bytes)?;
+            Ok(String::from_utf8(bytes)?)
+        } else {
+            let mut bytes: Vec<u8> = Vec::with_capacity(MAX_PREALLOCATED_CAPACITY);
+            let mut buf = [0u8; 64];
+            let mut read = 0;
+            while read < len {
+                let new = source.read(&mut buf)?;
+                if new == 0 {
+                    break;
+                } else {
+                    read += new;
+                    bytes.extend_from_slice(&buf[..new]);
+                }
+            }
+            if read == len {
+                Ok(String::from_utf8(bytes)?)
+            } else {
+                Err(ParseError {})
+            }
+        }
+    }
```

Comment:
Isn't this duplicated from the Vec<u8> one? I think we should reuse that if possible and then use `String::from_utf8`

Reply:
I had this strong sense of deja vu that I wrote this exact same code somewhere.

But it is just the factored out string indeed, not vec.